### PR TITLE
fix(runtime): terminate in-container process trees on timeout

### PR DESCRIPTION
**Pull request by dimileeh**
Automatically opened by AWF workspace `ws_ec22e84ff0ea4adc9c32c896` (agent: `codex`).

### Task
Implement a critical AWF runtime safety fix: when an agent or validation command launched through `docker compose exec` times out or is cancelled, AWF must not leave the in-container CLI process tree running while the executor moves on to validation, fix passes, push, or PR monitor work.

Observed dogfood failure:
- Claude Code hit an AWF idle timeout in workspace ws_8e0a2d21a0004223b3de6653.
- AWF killed the local `docker compose exec` client and recorded `AGENT_IDLE_TIMEOUT`.
- The original Claude process and child pytest were still alive inside the workspace container.
- AWF then moved into validation/fix-pass logic while the original agent could still mutate the worktree.

Expected behavior:
- Add a TDD-backed cleanup hook for timed-out/cancelled commands that execute inside a workspace container.
- For agent runs, AWF should tag or wrap the in-container command so timeout cleanup can reliably terminate the whole in-container process tree for that specific AWF invocation.
- It is acceptable to add an AWF-owned shell wrapper around the agent command, but do not depend on broad `pkill claude`/`pkill codex` that might kill unrelated debug processes.
- Apply the same safety to profile/validation commands launched via `docker compose exec` if they use the same command-runner path.
- After cleanup, AWF may continue salvage only if the in-container process tree was terminated or proven absent. If cleanup fails, fail the workspace with `infrastructure_failure` and an actionable reason code/message such as `EXEC_PROCESS_CLEANUP_FAILED` rather than running validation concurrently with a live timed-out agent.
- Operator cancellation/destroy must prevent subsequent validation fix passes from starting.
- Emit structured logs/events so the console/API make the root cause visible.

TDD requirements:
- Add focused failing tests first.
- Cover timeout of a docker-compose-exec agent command invokes targeted in-container cleanup.
- Cover cleanup failure maps to a clear infrastructure failure instead of validation/fix continuation.
- Cover cancellation/destroy status winning before a validation fix pass starts.
- Cover normal successful agent/validation runs do not invoke cleanup.
- Preserve existing tests and behavior for non-docker subprocesses.

Validation:
- uv run --python 3.12 --extra dev pytest tests/unit/common tests/unit/adapters tests/unit/control tests/unit/runtime -q
- uv run --python 3.12 --extra dev pytest --cov=awf --cov-report=term-missing
- uv run --python 3.12 --extra dev ruff check src/awf tests
- uv run --python 3.12 --extra dev mypy src/awf

Strict delivery rules:
- Follow AGENTS.md and `.awf/workspace.yml` planning/conformance contract.
- Do not reduce coverage thresholds or quality gates.
- Do not add fake tests, broad monkeypatches that skip behavior, xfail/skip shortcuts, or empty assertions.
- Keep changes scoped to the runtime command execution / adapter / executor cancellation safety surfaces unless your saved plan explains a necessary adjacent edit.
- Do not push, open PRs, merge, or switch branches. AWF owns branch lifecycle and PR monitoring.

---
Validation: 8 profile command(s) passed inside the workspace container.

### CODE REVIEW

File: src/awf/common/compose_exec.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    kill -"$1" "$pid" 2>/dev/null || true
+  done
+}}
+
+wait_until_absent() {{
+  i=0
+  while [ "$i" -lt 20 ]; do
+    collect_pids
+    [ -z "$pids" ] && return 0
+    sleep 0.1 2>/dev/null || sleep 1
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/common/compose_exec.py]: fixed in commit 3ff585e0945af31e58f49388299b0a29b8ee3b4e

File: src/awf/adapters/base.py [hunk 2]
```diff
@@ -223,33 +222,49 @@ async def run(
...
-                    agent=self.name.value,
-                    compose_project=compose_project,
-                    workspace_id=workspace_id,
-                    reason="runner does not support run_streaming",
-                )
-                result = await self._runner.run(args, input_bytes=b"")
-                if sinks is not None:
-                    await sinks.write_stdout(result.stdout)
-                    await sinks.write_stderr(result.stderr)
+                raise
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/adapters/base.py:254]: fixed in commit 425857fe75b060062980eee74f63219f79472d75

File: src/awf/runtime/validation.py [hunk 3]
```diff
@@ -465,45 +469,61 @@ async def _exec(
...
-            result = CommandResult(
-                returncode=124,
-                stdout="",
-                stderr=f"command timed out after {timeout_seconds}s",
-            )
-            reason_code = "PHASE_TIMEOUT"
-            if sinks is not None:
-                await sinks.write_stderr(result.stderr)
+                )
+                raise
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/runtime/validation.py:518]: FALSE POSITIVE: validation already calls cleanup_compose_exec_invocation_after_cancellation, which re-awaits the shielded cleanup task across repeated cancellations and surfaces any cleanup failure instead of orphaning it.
